contracts/externals/initializable.sol

```diff
@@ -0,0 +1,62 @@
+pragma solidity >=0.4.24 <0.7.0;
+
+
+/**
+ * @title Initializable
+ *
+ * @dev Helper contract to support initializer functions. To use it, replace
+ * the constructor with a function that has the `initializer` modifier.
+ * WARNING: Unlike constructors, initializer functions must be manually
+ * invoked. This applies both to deploying an Initializable contract, as well
+ * as extending an Initializable contract via inheritance.
+ * WARNING: When used with inheritance, manual care must be taken to not invoke
+ * a parent initializer twice, or ensure that all initializers are idempotent,
```

Comment:
I guess this is something we should get to the bottom of. Well spotted. 

Naively i would say yes, they are, but worth going through it all to confirm.

Reply:
So the pattern we are following is one external initializer (wallet) and the rest are internal (my assumption is that they can never be called again unless it is explicitly done via the inheriting contract.

Reply:
OK, so i have been through these constructors. 

``` ~/go/src/github.com/tokencard/contracts/contracts/internals   upgradeable-zeppelin ⍟2  sort -u /tmp/lame
_initializeControllable(_controllerNode_);
_initializeENSResolvable(_ens_);
_initializeGasTopUpLimit();
_initializeLoadLimit(_tokenWhitelistNode_);
_initializeOwnable(_ownerAddress_, false);
_initializeOwnable(_owner_, _transferable_);
_initializeSpendLimit(_spendLimit_);
_initializeTokenWhitelistable(_tokenWhitelistNode);
_initializeTokenWhitelistable(_tokenWhitelistNode_);```

I have been through all of those methods to check if they are idempotent :

This initialise method isn’t idempotent, from the Wallet, it has a time based component insofar as “now”:

```
    function _initializeLoadLimit(bytes32 _tokenWhitelistNode) internal {
        _initializeTokenWhitelistable(_tokenWhitelistNode);
        (, uint256 stablecoinMagnitude, , , , , ) = _getStablecoinInfo();
        require(stablecoinMagnitude > 0, "no stablecoin");
        _maximumLoadLimit = _MAXIMUM_STABLECOIN_LOAD_LIMIT * stablecoinMagnitude;
        _loadLimit = DailyLimitTrait.DailyLimit(_maximumLoadLimit, _maximumLoadLimit, now, 0, false);
    }
```

Reply:
I am guessing initialise functions with "now" in them aren't good, but i will leave you guys to figure this out :)

Reply:
I guess the whole confusion comes from the term itself. The comment is :
`When used with inheritance, manual care must be taken to not invoke a parent initializer twice, or ensure that all initializers are idempotent`

I don't exactly understand what _idempotent_ means in this context but my guess is that we shouldn't assume the strict mathematical definition here: if  an initializer can be called again with different parameters and change the state then it shouldn't be considered _idempotent_. That's why I followed the first approach ensuring that the initializers cannot be invoked twice. Please verify my statement, is ti safe to consider that internal functions cannot be invoked in any way unless explicitly stated in the code?

Reply:
I guess i will leave this to @i-stam and @nostdm to get to the bottom of.

Reply:
thanks for the note though. All i did is what i knew :)

Reply:
Non-Idempotent means that by invoking the method with the same parameters it will have different side effects based on whether it's being called the first, second or third time etc.

But I think this is actually handled by the `initializer` modifier from openzeppelin since it prevents the initializer being called multiple times.